server/platform/services/cache/redis.go

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package cache
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/v8/einterfaces"
+
+	"github.com/redis/rueidis"
+	"github.com/tinylib/msgp/msgp"
+	"github.com/vmihailenco/msgpack/v5"
+)
+
+const clientSideTTL = 5 * time.Minute
+
+type Redis struct {
+	name          string
+	client        rueidis.Client
+	defaultExpiry time.Duration
+	metrics       einterfaces.MetricsInterface
+}
+
+func NewRedis(opts *CacheOptions, client rueidis.Client) (*Redis, error) {
+	if opts.Name == "" {
+		return nil, errors.New("no name specified for cache")
+	}
+	return &Redis{
+		name:          opts.Name,
+		defaultExpiry: opts.DefaultExpiry,
+		client:        client,
+	}, nil
+}
+
+func (r *Redis) Purge() error {
+	// TODO: move to scan
+	keys, err := r.Keys()
+	if err != nil {
+		return err
+	}
+	return r.client.Do(context.Background(),
+		r.client.B().Del().
+			Key(keys...).
+			Build(),
+	).Error()
+}
+
+func (r *Redis) Set(key string, value any) error {
+	return r.SetWithExpiry(key, value, 0)
+}
+
+// SetWithDefaultExpiry adds the given key and value to the store with the default expiry. If
+// the key already exists, it will overwrite the previous value
+func (r *Redis) SetWithDefaultExpiry(key string, value any) error {
+	return r.SetWithExpiry(key, value, r.defaultExpiry)
+}
+
+// SetWithExpiry adds the given key and value to the cache with the given expiry. If the key
+// already exists, it will overwrite the previous value
+func (r *Redis) SetWithExpiry(key string, value any, ttl time.Duration) error {
+	now := time.Now()
+	defer func() {
+		if r.metrics != nil {
+			elapsed := time.Since(now).Seconds()
+			r.metrics.ObserveRedisEndpointDuration(r.name, "Set", elapsed)
+		}
+	}()
+	var buf []byte
+	var err error
+	// We use a fast path for hot structs.
+	if msgpVal, ok := value.(msgp.Marshaler); ok {
+		buf, err = msgpVal.MarshalMsg(nil)
```

Comment:
Random thought: have we considered using a pool of buffers to back this rather than allocating every time?

Reply:
We already use a pool in the `Get` path by using a pool of objects for some caches. Since the `Get` path is much more hit, rather than the `Set` path.

We could use a pool of buffers, but the challenge here is to identify exactly when to return the buffer back to pool.